### PR TITLE
Joystick on raspberry when joystick is enabled and sdl for windowing not

### PR DESCRIPTION
**Pull request by nadenislamarre**
This code doesn't handle device add/remove events.

Signed-off-by: Nicolas Adenis-Lamarre nicolas.adenis-lamarre@gmail.com

### CODE REVIEW

**Comment by wsnipex** [comment]
the configure change doesn't make much sense. SDL is only used for joystick, not for windowing (anymore).

**Comment by nadenislamarre** [comment]
this patch is initially for kodi 14.0.
Possibly the configure is not required anymore, but other files are required, while the current code still disable joystick for rpi and doesn't use sdl for joystick for rpi.

**Comment by Razzeee** [comment]
I think @garbear is working on joystick/pads etc right now

**Comment by garbear** [comment]
right, this code is all being dropped. It's all been moved to a binary add-ons. SDL isn't the default api anymore, we'll use the Linux Joystick API

**Comment by nadenislamarre** [comment]
hum, are you sure you will now use linux joystick api ? As far as i know, it's now something outdated, and udev must be used instead (which is what sdl does for example over linux).

**Comment by garbear** [comment]
I've confirmed the bug on RPi. @nadenislamarre, can you drop all changes outside xbmc/input/linux/LinuxInputDevices.cpp? I'll pull this commit into #8807 for testing.

File: xbmc/input/linux/LinuxInputDevices.cpp [hunk 1]
```diff
@@ -1103,6 +1103,23 @@ bool CLinuxInputDevices::CheckDevice(const char *device)
...
     return false;
 
+  // let others handle joysticks
+  unsigned long evbit[NBITS(EV_MAX)] = { 0 };
+  unsigned long keybit[NBITS(KEY_MAX)] = { 0 };
+  unsigned long absbit[NBITS(ABS_MAX)] = { 0 };
+
+  if (!((ioctl(fd, EVIOCGBIT(0, sizeof(evbit)), evbit) < 0) ||
+	(ioctl(fd, EVIOCGBIT(EV_KEY, sizeof(keybit)), keybit) < 0) ||
+	(ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(absbit)), absbit) < 0))) {
```
Comment (garbear) [inline xbmc/input/linux/LinuxInputDevices.cpp]: Our coding style is to put { on the next line. Can you change this please?

**Comment by maxnet** [comment]
This PR breaks touch-screen input devices.
Just that a device has absolute coordinates, does not make it automatically a joystick...

**Comment by nadenislamarre** [comment]
I've an update for this patch here :
based on udev ID_INPUT_JOYSTICK tag which is better than my previous patch
and will fix this issue.

patch for kodi 15 :
https://github.com/recalbox/recalbox-buildroot/blob/unified-4.0.0/package/kodi/0006-kodi-joystick-support.patch
​
And an other one in case you've not X11 or WINDOWS sdl event reader,
but it should be enclosed with the correct ifdef ; it allows you to
unplug/plug a device while kodi is running and catch de ADD/REMOVE events :

patch for kodi 15 :
https://github.com/recalbox/recalbox-buildroot/blob/unified-4.0.0/package/kodi/0007-kodi-joystick-plugandunplug-support.patch

**Comment by maxnet** [comment]
ID_INPUT_JOYSTICK is not perfect by default either, unless you are going to add extra rules that deal with the false positives.
https://github.com/denilsonsa/udev-joystick-blacklist

**Comment by nadenislamarre** [comment]
yes of course, but do you know a better suggestion ?
What is a joystick is a big question, sometimes not objectiv.

For example, the nintendo wiimote is splited in several linux events, one
for the buttons, one for the ir camera, one for the accelerometer and one
for each extension.
The buttons part is technically just a keyboard with 8 buttons.

The driver writter don't consider the wiimote as a joystick ; it depends of
what you do with it.
However, on the recalbox, i've added an udev rule (externally from the kodi
code) to make it a joystick, and it works.

I think that this id is a good way to externalize the information "is a
joystick" out from kodi.

Nicolas

**Comment by maxnet** [comment]
>yes of course, but do you know a better suggestion ?

User configurable instead of auto-detection?

**Comment by nadenislamarre** [comment]
Keyboards and joysticks are both linux events.
Ideally, one kodi module could handle all the events via libudev.
However, for the moment (at least in kodi 15, but as far as i understand,
tha'ts the same in 16),
there is one module specific for joysticks.
-- note : on recalbox, we modified the sdl lib too because of the
restrictions imposed for compatibility with the old linux joystick api
(udev should be used instead of sdl2 to avoid these limitations)
So, the question is : how to redirect the event to the correct module.

User configurable mean : you list all input events (so, you've a keyboard,
a joystick and a mouse, it's 3 items) in a screen, and you let the user
check/unckeck.
Ok. However, to initialize the default/unset value, ID_INPUT_JOYSTICK could
be combined while it works in most of the cases.

Note that joystick configuration required an xml too to map buttons.
So, having a configured joystick in that xml would mean : it's a joystick.

There are several solutions, i agree. But they require work to be done.
You can integrated my patch for the moment, and make something better later.

I did that patch for the recalbox project, and suggest it to kodi team, but
of course,
the decision is up to you to make differently.

**Comment by garbear** [comment]
With #8807, there is still a separate module for joysticks, but it has been moved to a binary add-on. What if we allow the binary add-on to scan for devices and grab exclusive access? We can use the blacklist above.

The add-on uses the linux joystick api, so it would have to be changed to udev, but that has been on my TODO list for a long time.

**Comment by nadenislamarre** [comment]
it looks nice ;-)

2016-02-15 18:24 GMT+01:00 Garrett Brown <notifications@github.com>:

**Comment by Razzeee** [comment]
So this PR is obsolete / broken as soon as @garbear work get's in?

**Comment by garbear** [comment]
merged nadenislamarre's alternative patch as part of #8897: ae41f305a583b61d58e3e4ba92e509306cef0481

**Comment by garbear** [comment]
Note that this fix only works if libudev is found at build time. OE builds for RPi have this, but standalone builds on Jenkins for RPi do not.

**Pull request closed without merging**